### PR TITLE
Implemented new decaying formula

### PR DESCRIPTION
**Pull request by juliolrmonteiro**
This is a solution for issue #174, reverting the rewards calculation from being based on current SEEDS exchange price to an exponential decaying formula.

### CODE REVIEW

**Comment by juliolrmonteiro** [comment]
The unit test currently doesn't verify the formula results, for the reward to actually start to decrease we need about 60 residents/citizens, any idea how can I add this many residents without making the test slow to a crawl? Maybe preset the test database?

File: src/seeds.accounts.cpp [hunk 1]
```diff
@@ -274,15 +283,28 @@ void accounts::refreward(name account, name new_status) {
   auto uitr = users.find(referrer.value);
   if (uitr != users.end()) {
     auto user_type = uitr->type;
+    auto num_users = is_citizen ? std::distance(citizens.cbegin(),citizens.cend()) : std::distance(residents.cbegin(),residents.cend());
```
Comment (n13) [inline src/seeds.accounts.cpp:287]: Can't use use distance... we need to keep track of the sizes of these tables in history

so history contract needs a sizes table that keeps track of resident and citizen table sizes - just like we keep track of sizes in other contracts.

Then we need an init method that initializes the size tables with the current values.

**Review by n13 — CHANGES_REQUESTED** [review]

File: src/seeds.accounts.cpp (hunk 1 above)
Comment (juliolrmonteiro) [reply src/seeds.accounts.cpp:291]: Done, please review again.

**Review by n13 — APPROVED** [review]

**Comment by n13** [comment]
Sorry this took so long didn't realize it was done

**Comment by n13** [comment]
Testet deploy log

❯ ./scripts/seeds.js run accounts history settings
Local
+  compiler command: eosio-cpp -abigen -I ./include -contract accounts -o ./artifacts/accounts.wasm ./src/seeds.accounts.cpp
+  accounts compiled
+  deploy accounts
+  creating accts.seeds
-  account accts.seeds already created
+  createAccount 
+  abi deployed
+  accounts code was already deployed
+  compiler command: eosio-cpp -abigen -I ./include -contract history -o ./artifacts/history.wasm ./src/seeds.history.cpp
+  history compiled
+  deploy history
+  creating histry.seeds
-  account histry.seeds already created
+  createAccount 
+  abi deployed
+  history code was already deployed
+  compiler command: eosio-cpp -abigen -I ./include -contract settings -o ./artifacts/settings.wasm ./src/seeds.settings.cpp
+  settings compiled
+  deploy settings
+  creating settgs.seeds
-  account settgs.seeds already created
+  createAccount 
+  abi deployed
+  settings code was already deployed

~/work/seeds/seeds-contracts master 27s
❯ cleos push action settgs.seeds reset '{}' -p settgs.seeds@active
executed transaction: a869deecde63cbef312fe0d862551bc12e5665f14348004d7a540b9476f449c9  96 bytes  4313 us
#  settgs.seeds <= settgs.seeds::reset          ""
warning: transaction executed locally, but may not be confirmed by the network yet         ] 

~/work/seeds/seeds-contracts master
❯ unlockcleoswallet.sh 
Error 3120007: Already unlocked
Error Details:
Wallet is already unlocked: default

~/work/seeds/seeds-contracts master
❯ cleos push action histry.seeds migrate '{}' -p histry.seeds@active
executed transaction: 8683988e9904e2cfbee98d4735534f6ae2c0742427f3ba967340657799ac9351  96 bytes  340 us
#  histry.seeds <= histry.seeds::migrate        ""
warning: transaction executed locally, but may not be confirmed by the network yet         ] 

~/work/seeds/seeds-contracts master
❯ ./scripts/seeds.js test accounts
Local
-  
    Local
  accounts

    reset proposals
    reset accounts
    reset token stats
    reset settings
    reset exchange
    reset escrow
    add users
    filling account with Seedds for bonuses [Change this]
    plant 50 seeds
    plant 100 seeds
    add referral firstuser is referrer for seconduser
    update reputation
    update user data
    make resident
    user not ready to become citizen{"code":500,"message":"Internal Service Error","error":{"code":3050003,"name":"eosio_assert_message_exception","what":"eosio_assert_message assertion failure","details":[{"message":"assertion failure with message: resident: user has less than required transactions number has: 0 needed: 1","file":"wasm_interface.cpp","line_number":1075,"method":"eosio_assert"}]}}
    test citizen    ✔  Given not citizen: should cant vote
    ✔  Given citizen: should can vote

    test resident    ✔  Given resident: should cant vote

    test citizen again
    test resident
    test citizen second user
    test testremove    ✔  Given referred became resudent: should reveive Seeds in escrow
    ✔  Given referred became citizen: should reveive Seeds in escrow
    ✔  Given changed reputation: should have correct values
    ✔  Given invited became resudent: should have community building score
    ✔  Given invited became citizen: should have community building score
    ✔  Given created user: should have correct timestamp
    ✔  Given invited user: should have row in table
    ✔  Given users table: should show joined users
    ✔  Given test-removed user: should have 1 fewer users than before

  vouching
    ✖  AssertionError [ERR_ASSERTION]: Missing ABI for account: harvst.seeds
    -------------------------------------------------------------------------
        operator: deepEqual
        diff: 
        source: at runMicrotasks (<anonymous>)
        at processTicksAndRejections (internal/process/task_queues.js:93:5)
        
        

  vouching with reputation
    ✖  AssertionError [ERR_ASSERTION]: Missing ABI for account: harvst.seeds
    -------------------------------------------------------------------------
        operator: deepEqual
        diff: 
        source: at runMicrotasks (<anonymous>)
        at processTicksAndRejections (internal/process/task_queues.js:93:5)
        
        

  Ambassador and Org rewards

    Ambassador and Org rewards =====    ✖  AssertionError [ERR_ASSERTION]: Missing ABI for account: orgs.seeds
    -----------------------------------------------------------------------
        operator: deepEqual
        diff: 
        source: at runMicrotasks (<anonymous>)
        at processTicksAndRejections (internal/process/task_queues.js:93:5)
        
        

  Proportional rewards
    ✖  AssertionError [ERR_ASSERTION]: Missing ABI for account: orgs.seeds
    -----------------------------------------------------------------------
        operator: deepEqual
        diff: 
        source: at runMicrotasks (<anonymous>)
        at processTicksAndRejections (internal/process/task_queues.js:93:5)
        
        

  make resident

    reset accounts
    reset history
    reset exchange
    reset token stats
    add users
    make resident - fail
    can resident - fail    ✔  Given does not fulfill criteria for resident: should be visitor
    ✔  Given does not fulfill criteria for resident - canresident is false: should be visitor

    plant 50 seeds
    make 10 transactions
    add referral
    update reputation
    can resident
    make resident    ✔  Given fulfills criteria for resident: should is resident

  make citizen
    ✖  AssertionError [ERR_ASSERTION]: Missing ABI for account: harvst.seeds
    -------------------------------------------------------------------------
        operator: deepEqual
        diff: 
        source: at runMicrotasks (<anonymous>)
        at processTicksAndRejections (internal/process/task_queues.js:93:5)
        
        

  reputation & cbs ranking

    reset accounts
    add users    ✔  Given 3 users with rep: should have entries in rep table
    ✔  Given 4 users with cbs: should have entries in cbs table
    ✔  Given cbs ranked in one go: should have same order
    ✔  Given 3 users with rep: should have number in sizes table
    ✔  Given 4 users with cbs: should have number in sizes table
    ✔  Given 4 users total: should have 4 in sizes table
    ✔  Given removed rep: should have entries in rep table
    ✔  Given removed rep from first user: should had 100, minus 2 is 98
    ✔  Given removed rep: should have number in sizes table



  passed: 24,  failed: 5  of 29 tests  (17.6s)

+  

~/work/seeds/seeds-contracts master 18s
❯ ./scripts/seeds.js run harvest
Local
+  compiler command: eosio-cpp -abigen -I ./include -contract harvest -o ./artifacts/harvest.wasm ./src/seeds.harvest.cpp
+  harvest compiled
+  deploy harvest
+  creating harvst.seeds
-  account harvst.seeds already created
+  createAccount 
+  abi deployed
+  harvest code was already deployed

~/work/seeds/seeds-contracts master 12s
❯ ./scripts/seeds.js test accounts                                  
Local
-  
    Local
  accounts

    reset proposals
    reset accounts
    reset token stats
    reset settings
    reset exchange
    reset escrow
    add users
    filling account with Seedds for bonuses [Change this]
    plant 50 seeds
    plant 100 seeds
    add referral firstuser is referrer for seconduser
    update reputation
    update user data
    make resident
    make citizen
    user not ready to become citizen{"code":500,"message":"Internal Service Error","error":{"code":3050003,"name":"eosio_assert_message_exception","what":"eosio_assert_message assertion failure","details":[{"message":"assertion failure with message: user has not referred enough residents or citizens: 0","file":"wasm_interface.cpp","line_number":1075,"method":"eosio_assert"}]}}
    test citizen    ✔  Given not citizen: should cant vote
    ✔  Given citizen: should can vote

    test resident    ✔  Given resident: should cant vote

    test citizen again
    test resident
    test citizen second user
    test testremove    ✔  Given referred became resudent: should reveive Seeds in escrow
    ✔  Given referred became citizen: should reveive Seeds in escrow
    ✔  Given changed reputation: should have correct values
    ✔  Given invited became resudent: should have community building score
    ✔  Given invited became citizen: should have community building score
    ✔  Given created user: should have correct timestamp
    ✔  Given invited user: should have row in table
    ✔  Given users table: should show joined users
    ✔  Given test-removed user: should have 1 fewer users than before

  vouching

    reset accounts
    reset harvest
    reset token stats
    reset settings
    reset exchange
    add users
    test citizen
    request vouch from user    ✔  Given init: should be empty XX

    vouch for user    ✔  Given after vouching: should get rep bonus for being vouched

    vouch for user but not resident or citizen
    test resident    ✔  Given after user is resident: should sponsor gets rep bonus for sponsoring resident
    ✔  Given resident vouch: should rep bonus
    ✔  Given after user is resident: should all sponsors gets rep bonus
    ✔  Given after user is citizen: should all sponsors gets rep bonus

    set max vouch to 3    ✔  Given max vouch reached: should can still vouch

    err expected {"code":500,"message":"Internal Service Error","error":{"code":3050003,"name":"eosio_assert_message_exception","what":"eosio_assert_message assertion failure","details":[{"message":"assertion failure with message: user is already fully vouched!","file":"wasm_interface.cpp","line_number":1075,"method":"eosio_assert"}]}}    ✔  Given vouch sponsor is not resident or citizen: should not be able to vouch
    ✔  Given vouch a second time: should not be able to vouch
    ✔  Given vouch past max points: should not be able to vouch

  vouching with reputation

    reset accounts
    reset harvest
    reset settings
    reset exchange
    add users
    test citizen    ✔  Given init: should be empty

    vouch for user    ✔  Given after vouching: should get rep bonus for being vouched

  Ambassador and Org rewards

    Ambassador and Org rewards =====    ✖  AssertionError [ERR_ASSERTION]: Missing ABI for account: orgs.seeds
    -----------------------------------------------------------------------
        operator: deepEqual
        diff: 
        source: at runMicrotasks (<anonymous>)
        at processTicksAndRejections (internal/process/task_queues.js:93:5)
        
        

  Proportional rewards
    ✖  AssertionError [ERR_ASSERTION]: Missing ABI for account: orgs.seeds
    -----------------------------------------------------------------------
        operator: deepEqual
        diff: 
        source: at runMicrotasks (<anonymous>)
        at processTicksAndRejections (internal/process/task_queues.js:93:5)
        
        

  make resident

    reset accounts
    reset history
    reset exchange
    reset token stats
    add users
    make resident - fail
    can resident - fail    ✔  Given does not fulfill criteria for resident: should be visitor
    ✔  Given does not fulfill criteria for resident - canresident is false: should be visitor

    plant 50 seeds
    make 10 transactions
    add referral
    update reputation
    can resident
    make resident    ✔  Given fulfills criteria for resident: should is resident

  make citizen

    reset settings
    reset harvest
    reset accounts
    reset token stats
    add users
    make citizen - fail
    can citizen - fail    ✔  Given does not fulfill criteria for citizen: should be visitor
    ✔  Given does not fulfill criteria for citizen - cancitizen false : should be false

    make citizen - fail    ✔  Given does not fulfill criteria for citizen: should be resident

    plant 200 seeds
    make 50 transaction
    add 3 referrals
    update reputation SCORE    ✔  Given does not fulfill criteria for citizen: should is resident still
    ✔  Given does not fulfill criteria for citizen: should is resident still

    can citizen - should succeed
    make citizen    ✔  Given does fulfill criteria for citizen: should is citizen

  reputation & cbs ranking

    reset accounts
    add users    ✔  Given 3 users with rep: should have entries in rep table
    ✔  Given 4 users with cbs: should have entries in cbs table
    ✔  Given cbs ranked in one go: should have same order
    ✔  Given 3 users with rep: should have number in sizes table
    ✔  Given 4 users with cbs: should have number in sizes table
    ✔  Given 4 users total: should have 4 in sizes table
    ✔  Given removed rep: should have entries in rep table
    ✔  Given removed rep from first user: should had 100, minus 2 is 98
    ✔  Given removed rep: should have number in sizes table



  passed: 42,  failed: 2  of 44 tests  (40s)

+  

~/work/seeds/seeds-contracts master 40s
❯ ./scripts/seeds.js run organizations
Local
+  compile failed for organizations error: Error: Contract not found: organizations No source file: ./src/seeds.organizations.cpp
+  error deploying 
+  TypeError: Cannot read property 'name' of undefined
    at deploy (/Users/nik/work/seeds/seeds-contracts/scripts/deploy.command.js:10:35)
    at async deployAction (/Users/nik/work/seeds/seeds-contracts/scripts/seeds.js:44:7)
    at async runAction (/Users/nik/work/seeds/seeds-contracts/scripts/seeds.js:85:3)
    at async batchCallFunc (/Users/nik/work/seeds/seeds-contracts/scripts/seeds.js:95:5)
    at async Command.<anonymous> (/Users/nik/work/seeds/seeds-contracts/scripts/seeds.js:138:5)

~/work/seeds/seeds-contracts master
❯ ./scripts/seeds.js run organization 
Local
+  compiler command: eosio-cpp -abigen -I ./include -contract organization -o ./artifacts/organization.wasm ./src/seeds.organization.cpp
+  organization compiled
+  deploy organization
+  creating orgs.seeds
-  account orgs.seeds already created
+  createAccount 
+  abi deployed
+  organization code was already deployed

~/work/seeds/seeds-contracts master 14s
❯ ./scripts/seeds.js test accounts    
Local
-  
    Local
  accounts

    reset proposals
    reset accounts
    reset token stats
    reset settings
    reset exchange
    reset escrow
    add users
    filling account with Seedds for bonuses [Change this]
    plant 50 seeds
    plant 100 seeds
    add referral firstuser is referrer for seconduser
    update reputation
    update user data
    make resident
    make citizen
    user not ready to become citizen{"code":500,"message":"Internal Service Error","error":{"code":3050003,"name":"eosio_assert_message_exception","what":"eosio_assert_message assertion failure","details":[{"message":"assertion failure with message: user has not referred enough residents or citizens: 0","file":"wasm_interface.cpp","line_number":1075,"method":"eosio_assert"}]}}
    test citizen    ✔  Given not citizen: should cant vote
    ✔  Given citizen: should can vote

    test resident    ✔  Given resident: should cant vote

    test citizen again
    test resident
    test citizen second user
    test testremove    ✔  Given referred became resudent: should reveive Seeds in escrow
    ✔  Given referred became citizen: should reveive Seeds in escrow
    ✔  Given changed reputation: should have correct values
    ✔  Given invited became resudent: should have community building score
    ✔  Given invited became citizen: should have community building score
    ✔  Given created user: should have correct timestamp
    ✔  Given invited user: should have row in table
    ✔  Given users table: should show joined users
    ✔  Given test-removed user: should have 1 fewer users than before

  vouching

    reset accounts
    reset harvest
    reset token stats
    reset settings
    reset exchange
    WARN: 10 attempts to find canonical signature
    add users
    test citizen
    request vouch from user    ✔  Given init: should be empty XX

    vouch for user    ✔  Given after vouching: should get rep bonus for being vouched

    vouch for user but not resident or citizen
    test resident    ✔  Given after user is resident: should sponsor gets rep bonus for sponsoring resident
    ✔  Given resident vouch: should rep bonus
    ✔  Given after user is resident: should all sponsors gets rep bonus
    ✔  Given after user is citizen: should all sponsors gets rep bonus

    set max vouch to 3    ✔  Given max vouch reached: should can still vouch

    err expected {"code":500,"message":"Internal Service Error","error":{"code":3050003,"name":"eosio_assert_message_exception","what":"eosio_assert_message assertion failure","details":[{"message":"assertion failure with message: user is already fully vouched!","file":"wasm_interface.cpp","line_number":1075,"method":"eosio_assert"}]}}    ✔  Given vouch sponsor is not resident or citizen: should not be able to vouch
    ✔  Given vouch a second time: should not be able to vouch
    ✔  Given vouch past max points: should not be able to vouch

  vouching with reputation

    reset accounts
    reset harvest
    reset settings
    reset exchange
    add users
    test citizen    ✔  Given init: should be empty

    vouch for user    ✔  Given after vouching: should get rep bonus for being vouched

  Ambassador and Org rewards

    Ambassador and Org rewards =====
    reset contracts
    add user
    ambassador invites a user named testorgowner
         Invite Secret: 354a41544c76474d674b6a4d6e4c4c48424864626a61707952376d595873594a
         Secret hash: a60bda0a4a8d0286b3a7d6ceea004a3e41e5a31c917a3d60af2851112cfce7dd
    token.seeds.transfer from seedsuseraaa to join.seeds (800.0000 SEEDS)
    join.seeds.invite from seedsuseraaa
    user accepts the invite and becomes a Seeds user
    join.seeds.accept invite for testorgowner
    accept success!
    user sends funds to orgs.seeds contract
    user creates org
    user sends funds to newly created org
    org signs up people
    org invites 2 users
         Invite Secret: 354b5470634a717a5757455938724c77476638705a5366464a574d3337775657
         Secret hash: fd0bb87177253851a6eebf7b3e817ac83e685c4615fc24ee3d8d351f226b9690
    token.seeds.transfer from testorg11111 to join.seeds (50.0000 SEEDS)
    join.seeds.invite from testorg11111
         Invite Secret: 354b6863507759373658464d484b796b58536b3555596b44444373525931386b
         Secret hash: 086f2827ee1006e1188a302575e774ed64ab74b104b940a01a9a16f7123b11de
    token.seeds.transfer from testorg11111 to join.seeds (100.0000 SEEDS)
    join.seeds.invite from testorg11111
    join.seeds.accept invite for seedsorgusr1
    accept success!
    join.seeds.accept invite for seedsorgusr2
    accept success!
    user 1 becomes resident    ✔  Given after resident: should reveive Seeds in escrow

    user 2 becomes citizen    ✔  Given after citizen 1: should reveive Seeds in escrow

    user 1 becomes citizen    ✔  Given after citizen 2: should reveive Seeds in escrow

  Proportional rewards

    reset contracts
    set exchange price
    add user
    individual invites a user named invited
         Invite Secret: 354a6b6f434b36324e6a4a35574c6473684456374b766e4b334e37476b6d4a6a
         Secret hash: 4944b67801c2cbbb4225c12964f81b125e4591c88f0680c294ad95314f6eb2b6
    token.seeds.transfer from seedsuseraaa to join.seeds (800.0000 SEEDS)
    join.seeds.invite from seedsuseraaa
    user accepts the invite and becomes a Seeds user
    join.seeds.accept invite for invited
    accept success!
    change settings
    user becomes resident    ✔  Given after resident: should receive Seeds in escrow

    user becomes citizen    ✔  Given after citizen: should receive Seeds in escrow

  make resident

    reset accounts
    reset history
    reset exchange
    reset token stats
    add users
    make resident - fail
    can resident - fail    ✔  Given does not fulfill criteria for resident: should be visitor
    ✔  Given does not fulfill criteria for resident - canresident is false: should be visitor

    plant 50 seeds
    make 10 transactions
    add referral
    update reputation
    can resident
    make resident    ✔  Given fulfills criteria for resident: should is resident

  make citizen

    reset settings
    reset harvest
    reset accounts
    reset token stats
    add users
    make citizen - fail
    can citizen - fail    ✔  Given does not fulfill criteria for citizen: should be visitor
    ✔  Given does not fulfill criteria for citizen - cancitizen false : should be false

    make citizen - fail    ✔  Given does not fulfill criteria for citizen: should be resident

    plant 200 seeds
    make 50 transaction
    add 3 referrals
    update reputation SCORE    ✔  Given does not fulfill criteria for citizen: should is resident still
    ✔  Given does not fulfill criteria for citizen: should is resident still

    can citizen - should succeed
    make citizen    ✔  Given does fulfill criteria for citizen: should is citizen

  reputation & cbs ranking

    reset accounts
    add users    ✔  Given 3 users with rep: should have entries in rep table
    ✔  Given 4 users with cbs: should have entries in cbs table
    ✔  Given cbs ranked in one go: should have same order
    ✔  Given 3 users with rep: should have number in sizes table
    ✔  Given 4 users with cbs: should have number in sizes table
    ✔  Given 4 users total: should have 4 in sizes table
    ✔  Given removed rep: should have entries in rep table
    ✔  Given removed rep from first user: should had 100, minus 2 is 98
    ✔  Given removed rep: should have number in sizes table


  passed: 47,  failed: 0  of 47 tests  (56s)

+  

~/work/seeds/seeds-contracts master 57s
❯ ./scripts/seeds.js test history

Local
-  
    Local
  make a transaction entry

    history reset
    accounts reset
    update status
    add transaction entry
    transactions result [
      {
        "id": 0,
        "to": "seedsuserbbb",
        "quantity": "10.0000 SEEDS",
        "timestamp": 1607411673
      }
    ]    ✔  Given transactions table: should have transaction entry

  make a history entry

    history reset
    history make entry
    now time 16074116
    check that history table has the entry
    timestamp 1607411674    ✔  Given action was tracked: should have table entry
    ✔  Given action was tracked: should timestamp is kinda close

    add resident
    add citizen    ✔  Given add resident entry: should have resident row
    ✔  Given add citizen entry: should have resident row

  org transaction entry

    history reset
    accounts reset
    settings reset
    org reset
    token reset weekly
    create org, users
    create organization
    add transaction entry from user to org    ✔  Given transactions table: should no change when org transact
    ✔  Given org tx table: should have 2 tx entry
    ✔  Given transactions table: should have transaction entry
    ✔  Given org transactions table: should have transaction entry
    ✔  Given org transactions table 2: should have transaction entries for both in and out


  passed: 10,  failed: 0  of 10 tests  (7s)

+  

~/work/seeds/seeds-contracts master 8s
❯ ./scripts/seeds.js test settings
Local
-  
    Local
  settings
    ✔  Given reset settings: should have config items with params
    ✔  Given reset settings: should have contract items with accounts
    ✔  Given set a value: should have value


  passed: 3,  failed: 0  of 3 tests  (812ms)

+  

~/work/seeds/seeds-contracts master
❯ EOSIO_NETWORK=telosTestnet ./scripts/seeds.js run accounts history settings

Telos Testnet
+  compiler command: eosio-cpp -abigen -I ./include -contract accounts -o ./artifacts/accounts.wasm ./src/seeds.accounts.cpp
+  accounts compiled
+  deploy accounts
+  creating accts.seeds
-  account accts.seeds already created
+  createAccount 
+  abi deployed
+  Success: accounts deployed to accts.seeds
+  accounts deployed
+  compiler command: eosio-cpp -abigen -I ./include -contract history -o ./artifacts/history.wasm ./src/seeds.history.cpp
+  history compiled
+  deploy history
+  creating histry.seeds
-  account histry.seeds already created
+  createAccount 
+  abi deployed
+  error deploying 
+  {"code":500,"message":"Internal Service Error","error":{"code":3080001,"name":"ram_usage_exceeded","what":"Account using more than allotted RAM usage","details":[{"message":"account histry.seeds has insufficient ram; needs 752020 bytes has 701398 bytes","file":"resource_limits.cpp","line_number":215,"method":"verify_account_ram_usage"}]}}
+  compiler command: eosio-cpp -abigen -I ./include -contract settings -o ./artifacts/settings.wasm ./src/seeds.settings.cpp
+  settings compiled
+  deploy settings
+  creating settgs.seeds
-  account settgs.seeds already created
+  createAccount 
+  abi deployed
+  Success: settings deployed to settgs.seeds
+  settings deployed

~/work/seeds/seeds-contracts master 1m 11s
❯ cleos -u https://test.hypha.earth/ system buyram -k illumination histry.seeds 1000 

Error 3120003: Locked wallet
Ensure that your wallet is unlocked before using it!
Error Details:
You don't have any unlocked wallet!

~/work/seeds/seeds-contracts master 6s
❯ unlockcleoswallet.sh 
Unlocked: default

~/work/seeds/seeds-contracts master
❯ cleos -u https://test.hypha.earth/ system buyram -k s33dst3stn3t histry.seeds 1000 

Error 3050003: eosio_assert_message assertion failure
Error Details:
assertion failure with message: overdrawn balance
pending console output: 

~/work/seeds/seeds-contracts master 7s
❯ cleos -u https://test.hypha.earth/ system buyram -k seeds histry.seeds 1000 

executed transaction: 43566f22527fdc34d17f8be2af214824fb5b7772915a4c45239b605b4ed06b96  112 bytes  1596 us
#         eosio <= eosio::buyrambytes           {"payer":"seeds","receiver":"histry.seeds","bytes":1024000}
#   eosio.token <= eosio.token::transfer        {"from":"seeds","to":"eosio.ram","quantity":"294.8069 TLOS","memo":"buy ram"}
#   eosio.token <= eosio.token::transfer        {"from":"seeds","to":"eosio.ramfee","quantity":"1.4815 TLOS","memo":"ram fee"}
#   eosio.token <= eosio.token::transfer        {"from":"eosio.ramfee","to":"eosio.rex","quantity":"1.4815 TLOS","memo":"transfer from eosio.ramfee ...
#         seeds <= eosio.token::transfer        {"from":"seeds","to":"eosio.ram","quantity":"294.8069 TLOS","memo":"buy ram"}
#     eosio.ram <= eosio.token::transfer        {"from":"seeds","to":"eosio.ram","quantity":"294.8069 TLOS","memo":"buy ram"}
#   eosio.trail <= eosio.token::transfer        {"from":"seeds","to":"eosio.ram","quantity":"294.8069 TLOS","memo":"buy ram"}
#         seeds <= eosio.token::transfer        {"from":"seeds","to":"eosio.ramfee","quantity":"1.4815 TLOS","memo":"ram fee"}
#  eosio.ramfee <= eosio.token::transfer        {"from":"seeds","to":"eosio.ramfee","quantity":"1.4815 TLOS","memo":"ram fee"}
#   eosio.trail <= eosio.token::transfer        {"from":"seeds","to":"eosio.ramfee","quantity":"1.4815 TLOS","memo":"ram fee"}
#  eosio.ramfee <= eosio.token::transfer        {"from":"eosio.ramfee","to":"eosio.rex","quantity":"1.4815 TLOS","memo":"transfer from eosio.ramfee ...
#     eosio.rex <= eosio.token::transfer        {"from":"eosio.ramfee","to":"eosio.rex","quantity":"1.4815 TLOS","memo":"transfer from eosio.ramfee ...
#   eosio.trail <= eosio.token::transfer        {"from":"eosio.ramfee","to":"eosio.rex","quantity":"1.4815 TLOS","memo":"transfer from eosio.ramfee ...
warning: transaction executed locally, but may not be confirmed by the network yet         ] 

~/work/seeds/seeds-contracts master 6s
❯ EOSIO_NETWORK=telosTestnet ./scripts/seeds.js run history                          

Telos Testnet
+  compiler command: eosio-cpp -abigen -I ./include -contract history -o ./artifacts/history.wasm ./src/seeds.history.cpp
+  history compiled
+  deploy history
+  creating histry.seeds
-  account histry.seeds already created
+  createAccount 
+  abi deployed
+  Success: history deployed to histry.seeds
+  history deployed

~/work/seeds/seeds-contracts master 28s
❯ cleos -u https://test.hypha.earth push action settgs.seeds reset '{}' -p settgs.seeds@active

executed transaction: e00304b14b09e2dbe4326ee8ce35546abd857d094d275a6dc9d0e6f1ae6d9e19  96 bytes  5475 us
#  settgs.seeds <= settgs.seeds::reset          ""
warning: transaction executed locally, but may not be confirmed by the network yet         ] 

~/work/seeds/seeds-contracts master 6s
❯ cleos -u https://test.hypha.earth get table histry.seeds histry.seeds sizes

{
  "rows": [],
  "more": false,
  "next_key": ""
}

~/work/seeds/seeds-contracts master
❯ cleos -u https://test.hypha.earth push action histry.seeds migrate '{}' -p histry.seeds@active

executed transaction: d38661b7f81a65cb934c2ea1999b4de2d94b70c473ac974074df72c22125da80  96 bytes  1189 us
#  histry.seeds <= histry.seeds::migrate        ""
warning: transaction executed locally, but may not be confirmed by the network yet         ] 

~/work/seeds/seeds-contracts master 8s
❯ cleos -u https://test.hypha.earth get table histry.seeds histry.seeds sizes                   

{
  "rows": [{
      "id": "citizens.sz",
      "size": 8
    },{
      "id": "regens.sz",
      "size": 0
    },{
      "id": "reptables.sz",
      "size": 0
    },{
      "id": "residents.sz",
      "size": 4
    }
  ],
  "more": false,
  "next_key": ""
}

~/work/seeds/seeds-contracts master
❯

**Comment by n13** [comment]
Mainnet deploy 

```

❯ EOSIO_NETWORK=telosMainnet ./scripts/seeds.js deploy settings

Telos Mainnet
+  deploy settings
+  creating settgs.seeds
-  account settgs.seeds already created
+  createAccount 
+  abi deployed
+  Success: settings deployed to settgs.seeds
+  settings deployed

~/work/seeds/seeds-contracts master 19s
❯ cleos -u https://node.hypha.earth push action settgs.seeds reset '{}' -p settgs.seeds@active

executed transaction: c7ff30a0eda2a6fbd9182205aa7eb2740b277e4825c0951c46024fae147c20fc  96 bytes  5544 us
#  settgs.seeds <= settgs.seeds::reset          ""
warning: transaction executed locally, but may not be confirmed by the network yet         ] 

~/work/seeds/seeds-contracts master
❯ EOSIO_NETWORK=telosMainnet ./scripts/seeds.js deploy accounts history

Telos Mainnet
+  deploy accounts
+  creating accts.seeds
-  account accts.seeds already created
+  createAccount 
+  abi deployed
+  Success: accounts deployed to accts.seeds
+  accounts deployed
+  deploy history
+  creating histry.seeds
-  account histry.seeds already created
+  createAccount 
+  abi deployed
+  Success: history deployed to histry.seeds
+  history deployed

~/work/seeds/seeds-contracts master 36s
❯ cleos -u https://node.hypha.earth get table histry.seeds histry.seeds sizes

{
  "rows": [],
  "more": false,
  "next_key": ""
}

~/work/seeds/seeds-contracts master
❯ cleos -u https://node.hypha.earth push action histry.seeds migrate '{}' -p histry.seeds@active

executed transaction: 846426f622fc59f28bf048a8ace7a38ab86403a517bc7e0e01342735dd4dccbe  96 bytes  4279 us
#  histry.seeds <= histry.seeds::migrate        ""
warning: transaction executed locally, but may not be confirmed by the network yet         ] 

~/work/seeds/seeds-contracts master 6s
❯ cleos -u https://node.hypha.earth get table histry.seeds histry.seeds sizes

{
  "rows": [{
      "id": "citizens.sz",
      "size": 70
    },{
      "id": "regens.sz",
      "size": 0
    },{
      "id": "reptables.sz",
      "size": 0
    },{
      "id": "residents.sz",
      "size": 143
    }
  ],
  "more": false,
  "next_key": ""
}

~/work/seeds/seeds-contracts master
❯ 
```

**Comment by n13** [comment]
@juliolrmonteiro I think the number of citizens is incorrect here - maybe history tables don't have a complete history of citizens (genesis citizens?)

Or were some demoted?

**Comment by n13** [comment]
Yes I think it's just because initial citizens aren't in that table

cleos -u https://node.hypha.earth get table histry.seeds histry.seeds citizens --lower illumination  
{
  "rows": [],
  "more": false,
  "next_key": ""
}

**Comment by n13** [comment]
In the long run, I don't think this discrepancy of 30 odd people makes a difference

What about demoted citizens @juliolrmonteiro ? Since we now demote them... I guess we should ignore that... which this table does, it doesn't update when citizens are demoted, these tables are meant to simply store the date someone became a citizen

**Pull request merged**